modules/listener/listener.bal

```diff
@@ -0,0 +1,154 @@
+    private json[] currentFileStatus = [];
+    private string specificFolderOrFileId;
+    private drive:Client driveClient;
+    private drive:WatchResponse watchResponse;
+    private boolean isWatchOnSpecificResource = false;
+    private boolean isAFolder = true;
+
+    # Listener initialization
+    public function init(ListenerConfiguration config) returns @tainted error? {
+        self.eventService = config.eventService;
+        self.httpListener = check new (config.port);
+        self.driveClient = check new (config.clientConfiguration);
+        if (config.specificFolderOrFileId is string) {
+            self.isAFolder = check checkMimeType(self.driveClient, config.specificFolderOrFileId.toString());
+        }
+        if (config.specificFolderOrFileId is string && self.isAFolder == true) {
+            check validateSpecificFolderExsistence(config.specificFolderOrFileId.toString(), self.driveClient);
+            self.specificFolderOrFileId = config.specificFolderOrFileId.toString();
+            self.watchResponse = check startWatch(config.callbackURL, self.driveClient, self.specificFolderOrFileId.
+            toString());
+            check getCurrentStatusOfDrive(self.driveClient, self.currentFileStatus, 
+            self.specificFolderOrFileId.toString());
+            self.isWatchOnSpecificResource = true;
+        } else if (config.specificFolderOrFileId is string && self.isAFolder == false) {
+            check validateSpecificFolderExsistence(config.specificFolderOrFileId.toString(), self.driveClient);
+            self.specificFolderOrFileId = config.specificFolderOrFileId.toString();
+            self.watchResponse = check startWatch(config.callbackURL, self.driveClient, self.specificFolderOrFileId);
+            self.isWatchOnSpecificResource = true;
+            check getCurrentStatusOfFile(self.driveClient, self.currentFileStatus, self.specificFolderOrFileId);
+        } else {
+            self.specificFolderOrFileId = EMPTY_STRING;
+            self.watchResponse = check startWatch(config.callbackURL, self.driveClient);
+            check getCurrentStatusOfDrive(self.driveClient, self.currentFileStatus);
+        }
+        self.channelUuid = self.watchResponse?.id.toString();
+        self.currentToken = self.watchResponse?.startPageToken.toString();
+        self.watchResourceId = self.watchResponse?.resourceId.toString();
+        log:print("Watch channel started in Google, id : " + self.channelUuid);
+    }
+    
+    public isolated function attach(http:Service s, string[]|string? name) returns error? {
+        return self.httpListener.attach(s, name);
+    }
+
+    public isolated function detach(http:Service s) returns error? {
+        return self.httpListener.detach(s);
+    }
+
+    public isolated function 'start() returns error? {
+        return self.httpListener.'start();
+    }
+
+    public isolated function gracefulStop() returns error? {
+        return ();
```

Comment:
Shouldn't we send a message to google API here stating that we are stopping? https://developers.google.com/calendar/v3/push#stopping-notifications

Reply:
Hi 
Yeah.. Handled stopWatch() in stop methods.
`public isolated function detach(service object {} s) returns error? {
        check stopWatchChannel(self.driveClient, self.channelUuid, self.watchResourceId);
        log:printDebug("Unsubscribed from the watch channel ID : " + self.channelUuid);
        return self.httpListener.detach(s);
    }

    public isolated function gracefulStop() returns error? {
        check stopWatchChannel(self.driveClient, self.channelUuid, self.watchResourceId);
        log:printDebug("Unsubscribed from the watch channel ID : " + self.channelUuid);
        return self.httpListener.gracefulStop();
    }

    public isolated function immediateStop() returns error? {
        check stopWatchChannel(self.driveClient, self.channelUuid, self.watchResourceId);
        log:printDebug("Unsubscribed from the watch channel ID : " + self.channelUuid);
        return self.httpListener.immediateStop();
    }`
Thanks